File: nim_core_macos/macos/Classes/common/NimResult.cpp
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

#include "NimResult.h"

using namespace flutter;

EncodableValue NimResult::getErrorResult(int code, const std::string& msg) {
  EncodableMap result;
  result.insert(std::make_pair(EncodableValue("code"), EncodableValue(code)));
  result.insert(
      std::make_pair(EncodableValue("errorDetails"), EncodableValue(msg)));
  result.insert(std::make_pair(EncodableValue("data"), EncodableValue()));
  return EncodableValue(result);
}

EncodableValue NimResult::getErrorResult(int code, const std::string& msg,
                                         const EncodableMap& data) {
  EncodableMap result;
  result.insert(std::make_pair(EncodableValue("code"), EncodableValue(code)));
  result.insert(
      std::make_pair(EncodableValue("errorDetails"), EncodableValue(msg)));
  result.insert(std::make_pair(EncodableValue("data"), EncodableValue(data)));
  return EncodableValue(result);
}

EncodableValue NimResult::getSuccessResult() {
  EncodableMap result;
  result.insert(std::make_pair(EncodableValue("code"), EncodableValue(0)));
  result.insert(
      std::make_pair(EncodableValue("errorDetails"), EncodableValue("")));
  result.insert(std::make_pair(EncodableValue("data"), EncodableValue()));
  return EncodableValue(result);
}

EncodableValue NimResult::getSuccessResult(int32_t data) {
  return getSuccessResult(EncodableValue(data));
}

EncodableValue NimResult::getSuccessResult(int64_t data) {
  return getSuccessResult(EncodableValue(data));
}

EncodableValue NimResult::getSuccessResult(const std::string& data) {
  return getSuccessResult(EncodableValue(data));
}

EncodableValue NimResult::getSuccessResult(const EncodableValue& data) {
  EncodableMap result;
  result.insert(std::make_pair(EncodableValue("code"), EncodableValue(0)));
  result.insert(
      std::make_pair(EncodableValue("errorDetails"), EncodableValue("")));
  result.insert(std::make_pair(EncodableValue("data"), data));
  return EncodableValue(result);
}

EncodableValue NimResult::getSuccessResult(const EncodableMap& data) {
  EncodableMap result;
  result.insert(std::make_pair(EncodableValue("code"), EncodableValue(0)));
  result.insert(
      std::make_pair(EncodableValue("errorDetails"), EncodableValue("")));
  result.insert(std::make_pair(EncodableValue("data"), data));
  return EncodableValue(result);
}

EncodableValue NimResult::getSuccessResult(const EncodableList& data) {
  EncodableMap result;
  result.insert(std::make_pair(EncodableValue("code"), EncodableValue(0)));
  result.insert(
      std::make_pair(EncodableValue("errorDetails"), EncodableValue("")));
  result.insert(std::make_pair(EncodableValue("data"), data));
  return EncodableValue(result);
}

EncodableValue NimResult::getSuccessResult(const std::string& msg,
                                           const EncodableMap& data) {
  EncodableMap result;
  result.insert(std::make_pair(EncodableValue("code"), EncodableValue(0)));
  result.insert(
      std::make_pair(EncodableValue("errorDetails"), EncodableValue(msg)));
  result.insert(std::make_pair(EncodableValue("data"), EncodableValue(data)));
  return EncodableValue(result);
}

File: nim_core/ios/Classes/NimResult.swift
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import Foundation
import NIMSDK

class NimResult {
  @objc var data: Any?
  @objc var code: NSNumber?
  @objc var errorDetails: String?

  init(_ data: Any?, _ code: NSNumber, _ msg: String?) {
    self.data = data
    self.code = code
    errorDetails = msg
  }

  func toDic() -> [String: Any] {
    var dic = [String: Any]()
    dic["code"] = code
    dic["errorDetails"] = errorDetails
    if let convert = data as? NimDataConvertProtrol {
      dic["data"] = convert.toDic()
    } else if let message = data as? NIMMessage {
      dic["data"] = message.toDic()
    } else if let object = data as? NSObject, let jsonData = object.yx_modelToJSONObject() {
      dic["data"] = jsonData
    } else if let tData = data {
      // 兼容 data 字段为基本类型
      dic["data"] = tData
    }
    return dic
  }

  class func error(_ msg: String) -> NimResult {
    let nimResult = NimResult(nil, NSNumber(integerLiteral: -1), msg)
    return nimResult
  }

  class func error(_ code: Int, _ msg: String) -> NimResult {
    let nimResult = NimResult(nil, NSNumber(integerLiteral: code), msg)
    return nimResult
  }

  class func success(_ msg: String) -> NimResult {
    let nimResult = NimResult(nil, NSNumber(integerLiteral: 0), msg)
    return nimResult
  }

  class func success(_ data: Any?) -> NimResult {
    let nimResult = NimResult(data, NSNumber(integerLiteral: 0), nil)
    return nimResult
  }

  class func successStringData(data: String) -> NimResult {
    let nimResult = NimResult(data, NSNumber(integerLiteral: 0), nil)
    return nimResult
  }

  class func success() -> NimResult {
    let nimResult = NimResult(nil, NSNumber(integerLiteral: 0), nil)
    return nimResult
  }

  class func success(_ data: Any?, _ msg: String) -> NimResult {
    let nimResult = NimResult(data, NSNumber(integerLiteral: 0), nil)
    return nimResult
  }
}
